File: object-oriented-programming-projects/project-5/project/rectangle.cpp
#include <iostream>
#include <iomanip>
#include <fstream>
#include <cmath>
#include <vector>

#include "myerrors.h"

using std::ostream;
using std::ofstream;
using std::vector;

#include "shape.h"

using SpaceShape::Shape;

#include "rectangle.h"

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::stod;

namespace SpaceRectangle
{
Rectangle::Rectangle():Rectangle(5,5){}
Rectangle::Rectangle(int myWidth, int myHeight)
	:width(myWidth),height(myHeight),x(0.0),y(0.0){
		if (width < 1) throw MyErrors(-3);
		if (height < 1) throw MyErrors(-4);
		totalArea += getArea();
		totalPerimeter += getPerimeter();
	}
Rectangle::Rectangle(int myWidth, int myHeight, double myX, double myY)
	:width(myWidth),height(myHeight),x(myX),y(myY){
		if (width < 1) throw MyErrors(-3);
		if (height < 1) throw MyErrors(-4);
		totalArea += getArea();
		totalPerimeter += getPerimeter();
	}
Rectangle::Rectangle(const Rectangle & other){
	width = other.width;
	height = other.height;
	x = other.x;
	y = other.y;
	totalArea += other.getArea();
	totalPerimeter += other.getPerimeter();
}
Rectangle & Rectangle::operator = (const Rectangle & other){
	totalArea -= getArea();
	totalPerimeter -= getPerimeter();
	width = other.width;
	height = other.height;
	x = other.x;
	y = other.y;
	totalArea += other.getArea();
	totalPerimeter += other.getPerimeter();
	return *this;
}
Rectangle::~Rectangle(){
	totalArea -= getArea();
	totalPerimeter -= getPerimeter();
}
void Rectangle::setWidth(int myWidth){
	totalArea -= getArea();
	totalPerimeter -= getPerimeter();
	width = myWidth;
	isWidthValid();
	totalArea += getArea();
	totalPerimeter += getPerimeter();
}
void Rectangle::setHeight(int myHeight){
	totalArea -= getArea();
	totalPerimeter -= getPerimeter();
	height = myHeight;
	isHeightValid();
	totalArea += getArea();
	totalPerimeter += getPerimeter();
}
void Rectangle::setX(double myX){x = myX;}
void Rectangle::setY(double myY){y = myY;}
int Rectangle::getWidth() const{return width;}
int Rectangle::getHeight() const{return height;}
double Rectangle::getPerimeter() const{return 2.0*width+2.0*height;}
double Rectangle::getArea() const{return width*height;}
double Rectangle::getX() const{return x;}
double Rectangle::getY() const{return y;}
double Rectangle::getTotalArea(){return totalArea;}
double Rectangle::getTotalPerimeter(){return totalPerimeter;}

Rectangle & Rectangle::operator ++ (){
	++x;
	++y;
	return *this;
}

Rectangle & Rectangle::operator ++ (int ignoreMe){
	Rectangle * temp = new Rectangle;
	*temp = *this;
	++x;
	++y;
	return *temp;
}

Rectangle & Rectangle::operator -- (){
	--x;
	--y;
	return *this;
}

Rectangle & Rectangle::operator -- (int ignoreMe){
	Rectangle * temp = new Rectangle;
	*temp = *this;
	--x;
	--y;
	return *temp;
}

const Rectangle Rectangle::operator + (double mySize) const{
	Rectangle temp = *this;
	temp.width = ((temp.width+mySize < 1) ? 1 : temp.width+mySize);
	temp.height = ((temp.height+mySize < 1) ? 1 : temp.height+mySize);
	return temp;
}

const Rectangle Rectangle::operator - (double mySize) const{
	Rectangle temp = *this;
	temp.width = ((temp.width-mySize < 1) ? 1 : temp.width-mySize);
	temp.height = ((temp.height-mySize < 1) ? 1 : temp.height-mySize);
	return temp;
}

const Rectangle operator + (double mySize, const Rectangle & myRectangle){
	Rectangle temp = myRectangle;
	temp.width = ((temp.width+mySize < 1) ? 1 : temp.width+mySize);
	temp.height = ((temp.height+mySize < 1) ? 1 : temp.height+mySize);
	return temp;
}

const Rectangle operator - (double mySize, const Rectangle & myRectangle){
	Rectangle temp = myRectangle;
	temp.width = ((temp.width-mySize < 1) ? 1 : temp.width-mySize);
	temp.height = ((temp.height-mySize < 1) ? 1 : temp.height-mySize);
	return temp;
}

ostream & Rectangle::printOut(ostream & print) const{
	print << "Your rectangle has a width of " << getWidth() << " and has a height of " 
		  << getHeight() << ", has a perimeter length of "
		  << static_cast <int> (getPerimeter()) << " and the area is " << getArea() << "." << endl;
	return print;
}

ofstream & Rectangle::draw(ofstream & file) const{
	file.open("rectangle.svg");
	file << R"(<svg version="1.1" baseProfile="full" width=")" << getWidth() << R"(" height=")" << getHeight() << R"(" xmlns="http://www.w3.org/2000/svg">)" << endl
		 << '\t' << R"(<rect width="100%" height="100%" fill="gray" />)" << endl;
	file << '\t' << R"(<rect x = ")" << 0 << R"(" y = ")" << 0 << R"(" width = ")" << getWidth() << R"(" height = ")" << getHeight() << R"(" fill = "purple" stroke = "black" stroke-width = "1" />)" << endl;
	file << "</svg>";
	file.close();
	return file;
}

ofstream & Rectangle::drawContainer(ofstream & file) const{
	file << '\t' << R"(<rect x = ")" << 0 << R"(" y = ")" << 0 << R"(" width = ")" << getWidth() << R"(" height = ")" << getHeight() << R"(" fill = "pink" stroke = "black" stroke-width = "1" />)" << endl;
	return file;
}

ofstream & Rectangle::drawSmall(ofstream & file) const{
	file << '\t' << R"(<rect x = ")" << getX() << R"(" y = ")" << getY() << R"(" width = ")" << getWidth() << R"(" height = ")" << getHeight() << R"(" fill = "purple" stroke = "black" stroke-width = "1" />)" << endl;
	return file;
}

void Rectangle::isWidthValid(){
	if (getWidth() < 1){
		cout << "Please enter a valid input for the width" << endl;
		string str;
		getline(cin,str,'\n');
		int size = str.size();
		if(size == 0){
			setWidth(-1);
			return;
		}
		if(size == 1 && (str[0] > '9' || str[0] < '0')){
			setWidth(-1);
			return;
		}
		for (int i = 0; i < size; ++i){
			if((str[i] > '9' || str[i] < '0')&&(str[i] != '.')){
				setWidth(-1);
				return;
			}
		}
		int count = 0;
		for (int i = 0; i < size; ++i){
			if(str[i] == '.')
				++count;
			if(count > 1){
				setWidth(-1);
				return;
			}
		}
		int length = stod(str);
		if (length < 1){
			setWidth(-1);
			return;
		}
		setWidth(length);
	}
}

void Rectangle::isHeightValid(){
	if (getHeight() < 1){
		cout << "Please enter a valid input for the height" << endl;
		string str;
		getline(cin,str,'\n');
		int size = str.size();
		if(size == 0){
			setHeight(-1);
			return;
		}
		if(size == 1 && (str[0] > '9' || str[0] < '0')){
			setHeight(-1);
			return;
		}
		for (int i = 0; i < size; ++i){
			if((str[i] > '9' || str[i] < '0')&&(str[i] != '.')){
				setHeight(-1);
				return;
			}
		}
		int count = 0;
		for (int i = 0; i < size; ++i){
			if(str[i] == '.')
				++count;
			if(count > 1){
				setHeight(-1);
				return;
			}
		}
		int length = stod(str);
		if (length < 1){
			setHeight(-1);
			return;
		}
		setHeight(length);
	}
}
}

File: object-oriented-programming-projects/project-7/project/Rectangle.java
import java.awt.Color;
import java.awt.Graphics;

/**
 * Rectangle is a Shape class that holds rectangle object
 * 
 * @author Tayfun
 *
 */
public class Rectangle implements Shape{

	private int width;
	private int height;
	private double x;
	private double y;
	private static double totalArea;
	private static double totalPerimeter;
	
	public Rectangle() throws MyErrors{
		this(5,5);
	}
	
	public Rectangle(int myWidth, int myHeight) throws MyErrors{
		this(myWidth,myHeight,0.0,0.0);
	}
	
	public Rectangle(int myWidth, int myHeight, double myX, double myY) throws MyErrors{
		x = myX;
		y = myY;
		width = myWidth;
		height = myHeight;
		if (width < 1) throw new MyErrors(-3);
		if (height < 1) throw new MyErrors(-4);
		totalArea += area();
		totalPerimeter += perimeter();
	}
	
	void setWidth(int myWidth) throws MyErrors{
		totalArea -= area();
		totalPerimeter -= perimeter();
		width = myWidth;
		if (width < 1)
			throw new MyErrors(-3);
		totalArea += area();
		totalPerimeter += perimeter();
	}
	void setHeight(int myHeight) throws MyErrors{
		totalArea -= area();
		totalPerimeter -= perimeter();
		height = myHeight;
		if (height < 1)
			throw new MyErrors(-4);
		totalArea += area();
		totalPerimeter += perimeter();
	}
	
	public Rectangle add(double d) throws MyErrors{
		Rectangle r = new Rectangle();
		r.setWidth(getWidth() + (int)d);
		r.setHeight(getHeight() + (int)d);
		return r;
	}
	
	public void setX(double myX){x = myX;}
	public void setY(double myY){y = myY;}
	public int getWidth() {return width;}
	public int getHeight() {return height;}
	public double perimeter() {return 2.0*width+2.0*height;}
	public double area() {return width*height;}
	public double getX() {return x;}
	public double getY() {return y;}
	public static double totalArea(){return totalArea;}
	public static double totalPerimeter(){return totalPerimeter;}
	
	public Rectangle increment(){
		++x;
		++y;
		return this;
	}
	public Rectangle decrement(){
		--x;
		--y;
		return this;
	}
	/**
	 * 
	 * draw draws the indicated shape on the screen
	 * its called from paintComponent method
	 * 
	 * @param g Graphics object passed from paintComponent method
	 * 
	 */
	public void draw(Graphics g){
		g.setColor(Color.MAGENTA);
		g.fillRect((int)x,(int)y,width,height);
		g.setColor(Color.BLACK);
		g.drawRect((int)x,(int)y,width,height);
	}
	
	public void print(){
		System.out.printf("Your rectangle has a width of %d and has a height of %d, has a perimeter length of %.2f and the area is %.2f.\n", getWidth(),getHeight(),perimeter(),area());
	}
	/**
	 * compareTo method compares two shape objects according to their areas
	 * 
	 * @param s another shape object to be compared
	 * @return an integer due to the result (<s -1, =s 0, >s 1)
	 */
	public int compareTo(Shape s){
		if(area()-s.area()<0.0001 && area()-s.area()>-0.0001)
			return 0;
		else if(area()-s.area()<0.0)
			return -1;
		else
			return 1;
	}
}
